File: sem5/jftt/list3/task2/field_operations.py
def field_add(a: int, b: int, field_characteristic: int) -> int:
    return (a + b) % field_characteristic


def field_sub(a: int, b: int, field_characteristic: int) -> int:
    return (a - b + field_characteristic) % field_characteristic


def field_mul(a: int, b: int, field_characteristic: int) -> int:
    return (a * b) % field_characteristic


def field_div(a: int, b: int, field_characteristic: int) -> int:
    return (a * field_pow(b, field_characteristic - 2, field_characteristic)) % field_characteristic


def field_pow(a: int, b: int, field_characteristic: int) -> int:
    result = 1
    while b > 0:
        if b % 2 == 1:
            result = field_mul(result, a, field_characteristic)
        a = field_mul(a, a, field_characteristic)
        b //= 2
    return result

File: sem5/jftt/list3/task1/src/fieldOperations.cpp
#include "fieldOperations.hpp"

long long fieldAdd(long long a, long long b, long long fieldCharacteristic) {
    return (a + b) % fieldCharacteristic;
}

long long fieldSub(long long a, long long b, long long fieldCharacteristic) {
    return (a - b + fieldCharacteristic) % fieldCharacteristic;
}

long long fieldMul(long long a, long long b, long long fieldCharacteristic) {
    return (a * b) % fieldCharacteristic;
}

long long fieldDiv(long long a, long long b, long long fieldCharacteristic) {
    return (a * fieldPow(b, fieldCharacteristic - 2, fieldCharacteristic)) % fieldCharacteristic;
}

long long fieldPow(long long a, long long b, long long fieldCharacteristic) {
    long long result = 1;
    while (b > 0) {
        if (b % 2 == 1) {
            result = fieldMul(result, a, fieldCharacteristic);
        }
        a = fieldMul(a, a, fieldCharacteristic);
        b /= 2;
    }
    return result;
}
